trino/auth.py

```diff
@@ -142,55 +143,40 @@ class _OAuth2TokenBearer(AuthBase):
     """
     MAX_OAUTH_ATTEMPTS = 5
 
-    class _AuthStep(Enum):
-        GET_REDIRECT_SERVER = 0
-        GET_TOKEN = 1
-
     def __init__(self, http_session, redirect_auth_url_handler=handle_redirect_auth_url):
         self._redirect_auth_url = redirect_auth_url_handler
-        self.http_session = http_session
         self._thread_local = threading.local()
+        self._pattern = re.compile(r'Bearer x_redirect_server="(.+)", x_token_server="(.+)"')
+        http_session.hooks['response'].append(self.__authenticate)
 
     def __call__(self, r):
-        if not hasattr(self._thread_local, 'init'):
-            self._thread_local.init = True
-            self._thread_local.token = None
-            self._thread_local.token_server = None
-            self._thread_local.attempts = 0
-            self._thread_local.auth_step = self._AuthStep.GET_REDIRECT_SERVER
-
-        if self._thread_local.token:
+        if hasattr(self._thread_local, 'token') and self._thread_local.token:
             r.headers['Authorization'] = "Bearer " + self._thread_local.token
 
         r.register_hook('response', self.__authenticate)
 
         return r
 
-    def __authenticate(self, r, **kwargs):
-        if (self._thread_local.auth_step == self._AuthStep.GET_REDIRECT_SERVER):
-            self.__process_get_redirect_server(r)
-        elif (self._thread_local.auth_step == self._AuthStep.GET_TOKEN):
-            self.__process_get_token(r)
-        return r
-
-    def __process_get_redirect_server(self, r):
-        if not 400 <= r.status_code < 500:
-            return r
+    def __authenticate(self, response, **kwargs):
+        if not 400 <= response.status_code < 500:
+            return response
 
         # we have to handle the authentication, may be token the token expired or it wasn't there at all
-        auth_info = r.headers.get('WWW-Authenticate')
+        auth_info = response.headers.get('WWW-Authenticate')
         if not auth_info:
             raise exceptions.TrinoAuthError("Error: header WWW-Authenticate not available in the response.")
 
-        matches = dict((s.strip().split('=', 1) for s in auth_info.split(",")))
-        if matches is None:
+        parts = list((s.strip().split('=', 1) for s in auth_info.split(",")))
+        if len(parts) != 2 or len(parts[0]) != 2 or len(parts[1]) != 2:
             raise exceptions.TrinoAuthError(f"Error: header info didn't match {auth_info}")
 
+        matches = dict(parts)
+
         auth_server = matches.get('Bearer x_redirect_server')
         if auth_server is None:
             raise exceptions.TrinoAuthError("Error: header info didn't have Bearer x_redirect_server")
 
-        token_server = matches.get('x_token_server')
+        token_server = matches.get('x_token_server').strip('"')
```

Comment:
Does this need to be done to `auth_server` too?

Reply:
Good catch! It was redundant